components/crypto_dot_com/browser/crypto_dot_com_json_parser.cc

```diff
@@ -0,0 +1,270 @@
+  base::Optional<base::Value>& records_v = value_with_error.value;
+
+  if (!records_v) {
+    LOG(ERROR) << "Invalid response, could not parse JSON, JSON is: " << json;
+    return false;
+  }
+
+  const base::Value* response = records_v->FindKey("response");
+  if (!response) {
+    return false;
+  }
+
+  const base::Value* result = response->FindKey("result");
+  if (!result) {
+    return false;
+  }
+
+  const base::Value* instruments = result->FindKey("instruments");
+  if (!instruments || !instruments->is_list()) {
+    return false;
+  }
+
+  for (const base::Value &instrument : instruments->GetList()) {
+    std::map<std::string, std::string> instrument_data;
+    const base::Value* pair = instrument.FindKey("instrument_name");
+    const base::Value* quote = instrument.FindKey("quote_currency");
+    const base::Value* base = instrument.FindKey("base_currency");
+
+    if (!(pair && pair->is_string()) ||
+        !(quote && quote->is_string()) ||
+        !(base && base->is_string())) {
+      continue;
+    }
+
+    instrument_data.insert({"pair", pair->GetString()});
+    instrument_data.insert({"quote", quote->GetString()});
+    instrument_data.insert({"base", base->GetString()});
+
+    pairs->push_back(instrument_data);
+  }
+
+  return true;
+}
+
+bool CryptoDotComJSONParser::GetRankingsFromJSON(
+    const std::string& json,
+    CryptoDotComAssetRankings* rankings) {
+  if (!rankings) {
+    return false;
+  }
+
+  base::JSONReader::ValueWithError value_with_error =
+      base::JSONReader::ReadAndReturnValueWithError(
+          json, base::JSONParserOptions::JSON_PARSE_RFC);
+  base::Optional<base::Value>& records_v = value_with_error.value;
+
+  if (!records_v) {
+    LOG(ERROR) << "Invalid response, could not parse JSON, JSON is: " << json;
+    return false;
+  }
+
+  const base::Value* response = records_v->FindKey("response");
+  if (!response) {
+    return false;
+  }
+
+  const base::Value* result = response->FindKey("result");
+  if (!result) {
+    return false;
+  }
+
+  std::vector<std::map<std::string, std::string>> gainers;
+  std::vector<std::map<std::string, std::string>> losers;
+
+  // Both gainers and losers are part of the "gainers" list
+  const base::Value* rankings_list = result->FindKey("gainers");
+  if (!rankings_list || !rankings_list->is_list()) {
+    // Gainers and losers should return empty on a bad response
+    rankings->insert({"gainers", gainers});
+    rankings->insert({"losers", losers});
+    return false;
+  }
+
+  for (const base::Value &ranking : rankings_list->GetList()) {
+    std::map<std::string, std::string> ranking_data;
+    const base::Value* pair = ranking.FindKey("instrument_name");
+    const base::Value* change = ranking.FindKey("percent_change");
+    const base::Value* last = ranking.FindKey("last_price");
+
+    if (!(pair && pair->is_string()) ||
+        !(change && change->is_string()) ||
+        !(last && last->is_string())) {
```

Comment:
nit: could be flattened out with De Morgan's law

Reply:
fixed